packages/web3/src/contract/dapp-tx-builder.ts

```diff
@@ -0,0 +1,210 @@
+/*
+Copyright 2018 - 2022 The Alephium Authors
+This file is part of the alephium project.
+
+The library is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+The library is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with the library. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { contractIdFromAddress } from '../address'
+import { Val } from '../api'
+import {
+  AddressConst,
+  ApproveAlph,
+  ApproveToken,
+  BytesConst,
+  CallExternal,
+  ConstFalse,
+  ConstTrue,
+  Dup,
+  Instr,
+  Pop,
+  toI256,
+  toU256,
+  U256Const,
+  Method
+} from '../codec'
+import { LockupScript, lockupScriptCodec } from '../codec/lockup-script-codec'
+import { scriptCodec } from '../codec/script-codec'
+import { ALPH_TOKEN_ID } from '../constants'
+import { TraceableError } from '../error'
+import { SignExecuteScriptTxParams } from '../signer'
+import { base58ToBytes, binToHex, HexString, hexToBinUnsafe, isBase58, isHexString } from '../utils'
+
+export class DappTransactionBuilder {
+  private callerLockupScript: LockupScript
+  private approvedAssets: Map<HexString, bigint>
+  private instrs: Instr[]
+
+  constructor(public readonly callerAddress: string) {
+    try {
+      this.callerLockupScript = lockupScriptCodec.decode(base58ToBytes(this.callerAddress))
+      if (this.callerLockupScript.kind !== 'P2PKH' && this.callerLockupScript.kind !== 'P2SH') {
+        throw new Error(`Expected a P2PKH address or P2SH address`)
+      }
+    } catch (error) {
+      throw new TraceableError(`Invalid caller address: ${callerAddress}`, error)
+    }
+    this.approvedAssets = new Map<HexString, bigint>()
+    this.instrs = []
+  }
+
+  callContract(params: {
+    contractAddress: string
+    methodIndex: number
+    args: Val[]
+    attoAlphAmount?: bigint
+    tokens?: { id: HexString; amount: bigint }[]
+    retLength?: number
+  }) {
+    if (!isBase58(params.contractAddress)) {
+      throw new Error(`Invalid contract address: ${params.contractAddress}, expected a base58 string`)
+    }
+    const contractLockupScript = lockupScriptCodec.decode(base58ToBytes(params.contractAddress))
+    if (contractLockupScript.kind !== 'P2C') {
+      throw new Error(`Invalid contract address: ${params.contractAddress}, expected a P2C address`)
+    }
```

Comment:
[Nit] Maybe we can check `contractAddress` using [isContractAddress](https://github.com/alephium/alephium-web3/blob/master/packages/web3/src/address/address.ts#L88)?